### PR TITLE
Resolve JMeter execute via TestGrid

### LINKED ISSUE #80: Execute JMeter programmatically
**Description:**
As of now, we are running the JMeter via calling the JMeter shell script directly from Java. This will avoid TestGrid from running on Windows machines.

### PR DESCRIPTION
**Pull request by vidurananayakkara**
## Purpose
Resolves #80 

## Goals
* JMeter should be executed via Java
* Increase unit test coverage in automation module

## Release note
* Fixed JMeter to execute via Java

## Automation tests
 - Unit tests 

|        Class        | Class coverage | Method coverage | Line coverage |
|:-------------------:|:--------------:|:---------------:|:-------------:|
| JMeterExecutor      |   100% (1/1)   |    100% (6/6)   |  85% (58/68)  |
| TestExecutorFactory |   100% (1/1)   |    100% (1/1)   |   80% (4/5)   |

 - Integration tests
Included herewith this PR

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        Path scriptFileName = Paths.get(script).getFileName();
+        if (scriptFileName == null) {
+            throw new TestAutomationException(StringUtil.concatStrings("Script file ", script, " cannot be located."));
+        }
+        String resultFile =
+                Paths.get(testLocation, "Results", "Jmeter", scriptFileName + ".xml").toAbsolutePath().toString();
+        ResultCollector resultCollector = new ResultCollector(summariser);
+
+        resultCollector.setFilename(resultFile);
+        testPlanTree.add(testPlanTree.getArray()[0], resultCollector);
```
Comment (kasunbg) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: Can we guarantee the first element will be available?

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Overrides the JMeter properties with the properties required for the current deployment.
+     *
+     * @param testLocation directory location of the tests
+     * @param testName     test name
+     * @param deployment   deployment details of the current pattern
+     */
+    private void overrideJMeterConfig(String testLocation, String testName, Deployment deployment) {
+        Path path = Paths.get(testLocation, "JMeter", testName, "src", "test", "resources", "user.properties");
```
Comment (kasunbg) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: Should we hard-code this?

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        if (Files.exists(path)) {
```
Comment (kasunbg) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: log if you cannot find it.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            JMeterUtils.loadJMeterProperties(path.toAbsolutePath().toString());
```
Comment (kasunbg) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: A utils class should never store state. Is it the case here?

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                    .toAbsolutePath();
+            setPermission(shellScriptPath);
+        }
+
+        // Start server
+        String execFile = SystemUtils.IS_OS_WINDOWS ?
+                          Paths.get(distributionLocation, "bin", "wso2server.bat").toAbsolutePath().toString() :
+                          Paths.get(distributionLocation, "bin", "wso2server.sh").toAbsolutePath().toString();
+        ProcessBuilder processBuilder = new ProcessBuilder(execFile);
+        processBuilder.start();
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: Unit tests should only do functional testing. If you want to test how the composed server functions, then we should write integration tests instead.

For your usecase, it seems you need a WSO2 server to verify jmeter. I would suggest that we mock the wso2server https endpoint. We can use Mockito.

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+        String testLocation = Paths.get(resource.getPath(), "SolutionPattern22", "Tests")
+                .toAbsolutePath().toString();
+        TestExecutor testExecutor = new JMeterExecutor();
+        testExecutor.init(testLocation, "solution22");
+        testExecutor.execute(testScript, deployment);
+
+        // Assert test result file
+        Assert.assertTrue(Files.exists(Paths.get(resource.getPath(), "SolutionPattern22", "Tests", "Results",
+                "Jmeter", "Create-Test-Users.jmx.xml")));
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: Is it enough to only verify the existence of the file?

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    public void testNoJMeterSaveConfig() throws TestAutomationException {
+        // Set JMeter home
+        EnvironmentUtil.setEnvironmentVariable(TestGridConstants.JMETER_HOME, "JMeter Home");
+
+        TestExecutor testExecutor = new JMeterExecutor();
+        testExecutor.init("Test Location", "solution22");
+        testExecutor.execute("", new Deployment());
+
+        // Unset JMeter home
+        EnvironmentUtil.unsetEnvironmentVariable(TestGridConstants.JMETER_HOME);
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: No assertions in the method.

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+                "src", "test", "jmeter", "nofile.jmx").toAbsolutePath().toString();
+
+        String testLocation = Paths.get(resource.getPath(), "SolutionPattern22", "Tests")
+                .toAbsolutePath().toString();
+        TestExecutor testExecutor = new JMeterExecutor();
+        testExecutor.init(testLocation, "solution22");
+        testExecutor.execute(testScript, new Deployment());
+
+        // Unset JMeter home
+        EnvironmentUtil.unsetEnvironmentVariable(TestGridConstants.JMETER_HOME);
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: No assertions.

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    }
+
+    /**
+     * Creates and returns a {@link Deployment} for the given name.
+     *
+     * @param name  deployment name
+     * @param hosts list of hosts for the deployment
+     * @return an instance of {@link Deployment}
+     */
+    private Deployment createDeployment(String name, List<Host> hosts) {
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: How #createDeploymentPojo instead?

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    }
+
+    /**
+     * Creates and returns a {@link Port} instance for the given port number and protocol.
+     *
+     * @param portNumber port number
+     * @param protocol   protocol of the port
+     * @return an instance of {@link Port}
+     */
+    private Port createPort(int portNumber, String protocol) {
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: Same as above.

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/util/ZipUtil.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+        // Iterates over entries in the zip file
+        while (entry != null) {
+            String filePath = destDirectory + File.separator + entry.getName();
+            if (!entry.isDirectory()) {
+                // If the entry is a file, extracts it
+                extractFile(zipIn, filePath);
+            } else {
+                // If the entry is a directory, make the directory
+                File dir = new File(filePath);
```
Comment (kasunbg) [inline automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/util/ZipUtil.java]: Shall we use the Paths API?

File: automation/src/test/resources/testng.xml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+<suite name="TestGrid - Test Automation Module">
+    <test name="executor-test" parallel="false">
+        <parameter name="jmeterHome" value="/opt/apache-jmeter-3.2"/>
+        <classes>
+            <class name="org.wso2.carbon.testgrid.automation.executor.JMeterExecutorTest"/>
+            <class name="org.wso2.carbon.testgrid.automation.executor.TestExecutorFactoryTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (kasunbg) [inline automation/src/test/resources/testng.xml]: Please add a new line at the end.

Comment (vidurananayakkara) [reply automation/src/test/resources/testng.xml]: New line added

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java (hunk 2 above)
Comment (vidurananayakkara) [reply automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: This should be discussed under https://github.com/wso2-incubator/wso2-test-grid/issues/90 immediately. After finalizing that we can proceed to do the changes. WDYT?

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java (hunk 4 above)
Comment (vidurananayakkara) [reply automation/src/main/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutor.java]: True. But this is the JMeters way of loading JMeter properties (required) and hence we have to follow it.

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java (hunk 5 above)
Comment (vidurananayakkara) [reply automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: Mocked via an external server (https://www.mocky.io/).

File: automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java (hunk 6 above)
Comment (vidurananayakkara) [reply automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java]: No.With issue https://github.com/wso2-incubator/wso2-test-grid/issues/84 we are taking the results out programmatically via JMeter and will be handled there.

**Comment by vidurananayakkara** [comment]
@kasunbg Please note that for test methods without assertions there is an expected exception and an expected exception message.

**Pull request merged**